### PR TITLE
Adding support for proto extensions

### PR DESCRIPTION
**Pull request by jeremiepiotte-districtm**
- Tested building schemas locally. ✅ 
- Tested using schemas to create BigQuery tables. ✅
- Code coverage ✅  
- That PR was a pain to figure out. 🤯 

[Ticket in context](https://districtm.atlassian.net/browse/DMX-1234?atlOrigin=eyJpIjoiNGQ4MmNiYWZhZGI1NDczM2E5ZjMxZTM4NDkzOTE1Y2QiLCJwIjoiaiJ9)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -413,6 +450,51 @@ func convert(req *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, e
...
+			}
+			nodeIdx += 1
+		}
+
+		// Not the right package or message
+		if nodeIdx == 0 || msg.GetName() != extendeeNodes[nodeIdx] {
+			continue
+		}
+
+		foundNode := (*descriptor.DescriptorProto)(nil)
```
Comment (QuentinMcgaw-districtm) [inline main.go:478]:
```suggestion
		var foundNode *descriptor.DescriptorProto
```

File: main.go [hunk 2]
```diff
@@ -413,6 +450,51 @@ func convert(req *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, e
...
+		nestedMsgs := msg.GetNestedType()
+		for i := nodeIdx; i < len(extendeeNodes); i += 1 {
+			foundNode = nil
```
Comment (QuentinMcgaw-districtm) [inline main.go:481]: you seem to reset it to nil and the break below only breaks the inner for loop, is this expected?

File: main.go [hunk 3]
```diff
@@ -413,6 +450,51 @@ func convert(req *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, e
...
+		for _, node := range pkgNodes {
+			if node != extendeeNodes[nodeIdx] {
+				nodeIdx = 0
+				break
+			}
+			nodeIdx += 1
+		}
+
+		// Not the right package or message
+		if nodeIdx == 0 || msg.GetName() != extendeeNodes[nodeIdx] {
```
Comment (QuentinMcgaw-districtm) [inline main.go:474]: Explicit `mismatch` for clarity, feel free to rename

```suggestion
		nodeIdx := 1
		mismatch := false
		for _, node := range pkgNodes {
			if node != extendeeNodes[nodeIdx] {
				mismatch = true
				break
			}
			nodeIdx += 1
		}

		// Not the right package or message
		if mismatch || msg.GetName() != extendeeNodes[nodeIdx] {
```

File: main.go [hunk 4]
```diff
@@ -413,6 +450,51 @@ func convert(req *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, e
...
+		}
+
+		// skipping leading "."
+		nodeIdx := 1
+		for _, node := range pkgNodes {
+			if node != extendeeNodes[nodeIdx] {
+				nodeIdx = 0
+				break
+			}
+			nodeIdx += 1
```
Comment (QuentinMcgaw-districtm) [inline main.go:470]: :older_woman: 

```suggestion
			nodeIdx++
```

File: main.go [hunk 5]
```diff
@@ -413,6 +450,51 @@ func convert(req *plugin.CodeGeneratorRequest) (*plugin.CodeGeneratorResponse, e
...
+
+		// We already know this extendee cannot be in this package.
+		if len(pkgNodes) > len(extendeeNodes) {
+			continue
+		}
+
+		// skipping leading "."
+		nodeIdx := 1
+		for _, node := range pkgNodes {
+			if node != extendeeNodes[nodeIdx] {
```
Comment (QuentinMcgaw-districtm) [inline main.go:466]: that will fail if there is no dot in `ext.GetExtendee()`. I would recommend to start at `nodeIdx := 0` and do a len check on extendeeNodes above:

```go
if len(extendeeNodes) < 1 {
    continue
}
extendeeNodes = extendeeNodes[1:]
```

File: main.go (hunk 4 above)
Comment (jeremiepiotte-districtm) [reply main.go:470]: What, I didn't know we could do that in Golang... Is that a new feature ?

File: main.go (hunk 2 above)
Comment (jeremiepiotte-districtm) [reply main.go:481]: Yes, it is expected behavior

File: main.go (hunk 5 above)
Comment (jeremiepiotte-districtm) [reply main.go:466]: Yes, I did that initially, but then I learned that it always has a dot.  But sure, I can revert back

File: main.go (hunk 4 above)
Comment (QuentinMcgaw-districtm) [reply main.go]: :smile: No. You should setup https://golangci-lint.run/usage/install/#local-installation there is a lot to learn from it

File: main.go (hunk 5 above)
Comment (QuentinMcgaw-districtm) [reply main.go:475]: As in, let's not trust whatever source it's coming from. If one day the `.` isn't there, we'll get panic all over the place haha

File: main.go (hunk 4 above)
Comment (jeremiepiotte-districtm) [reply main.go]: Installed

File: main.go [hunk 6]
```diff
@@ -460,22 +460,31 @@ func addExtensions(msg *descriptor.DescriptorProto, extensions []*descriptor.Fie
 			continue
 		}
 
+		if len(extendeeNodes) < 1 {
```
Comment (QuentinMcgaw-districtm) [inline main.go:463]: Actually, my bad, that should be 

```suggestion
		if len(extendeeNodes) < 2 {
```

as we'll have one element even without `.` or an empty string. As soon as there would be a `.` we're guaranteed to have at least 2 elements.

**Review by QuentinMcgaw-districtm — APPROVED** [review]

**Pull request merged**